src/librustc/traits/error_reporting.rs

```diff
@@ -135,65 +135,111 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
 
         let ity = tcx.lookup_item_type(did);
         let (tps, rps, _) =
-            (ity.generics.types.get_slice(subst::TypeSpace),
-             ity.generics.regions.get_slice(subst::TypeSpace),
+            (ity.generics.types.get_slice(TypeSpace),
+             ity.generics.regions.get_slice(TypeSpace),
              ity.ty);
 
         let rps = self.region_vars_for_defs(obligation.cause.span, rps);
         let mut substs = subst::Substs::new(
             subst::VecPerParamSpace::empty(),
             subst::VecPerParamSpace::new(rps, Vec::new(), Vec::new()));
         self.type_vars_for_defs(obligation.cause.span,
-                                subst::ParamSpace::TypeSpace,
+                                TypeSpace,
                                 &mut substs,
                                 tps);
         substs
     }
 
-    fn impl_with_self_type_of(&self,
-                              trait_ref: ty::PolyTraitRef<'tcx>,
-                              obligation: &PredicateObligation<'tcx>)
+    fn fuzzy_match_tys(&self, a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+        /// returns the fuzzy category of a given type, or None
+        /// if the type can be equated to any type.
+        fn type_category<'tcx>(t: Ty<'tcx>) -> Option<u32> {
+            match t.sty {
+                ty::TyBool => Some(0),
+                ty::TyChar => Some(1),
+                ty::TyStr => Some(2),
+                ty::TyInt(..) | ty::TyUint(..) |
+                ty::TyInfer(ty::IntVar(..)) => Some(3),
+                ty::TyFloat(..) | ty::TyInfer(ty::FloatVar(..)) => Some(4),
+                ty::TyEnum(..) => Some(5),
+                ty::TyStruct(..) => Some(6),
+                ty::TyBox(..) | ty::TyRef(..) | ty::TyRawPtr(..) => Some(7),
+                ty::TyArray(..) | ty::TySlice(..) => Some(8),
+                ty::TyFnDef(..) | ty::TyFnPtr(..) => Some(9),
+                ty::TyTrait(..) => Some(10),
+                ty::TyClosure(..) => Some(11),
+                ty::TyTuple(..) => Some(12),
+                ty::TyProjection(..) => Some(13),
+                ty::TyParam(..) => Some(14),
+                ty::TyInfer(..) | ty::TyError => None
+            }
+        }
+
+        match (type_category(a), type_category(b)) {
+            (Some(cat_a), Some(cat_b)) => match (&a.sty, &b.sty) {
+                (&ty::TyStruct(def_a, _), &ty::TyStruct(def_b, _)) |
+                (&ty::TyEnum(def_a, _), &ty::TyEnum(def_b, _)) =>
+                    def_a == def_b,
+                _ => cat_a == cat_b
+            },
+            // infer and error are
```

Comment:
Nit: incomplete sentence... "infer and error are fuzzy equal to all types", I guess?